### PR TITLE
Epmlstrtai-456

### PR DESCRIPTION
**Pull request by AfanasievYuriy**
(https://jirapct.epam.com/jira/browse/EPMLSTRTAI-456 - see story here)
Added project essence (controller, service, DTOs)
Changed request URLs for all controllers
Changed suit controller and service (interaction with Project was added)
Not all tests were added (still working on it)

### CODE REVIEW

File: src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    @NotNull
+    @Size(min = 1, max = 255)
+    private String name;
+
+    @Size(max = 255)
+    private String description;
+
+    private List<SuitDTO> suits;
+
+    private Set<UserDTO> users;
```
Comment (artemSpb96) [inline src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java]: Add valid annotations, please.

File: src/main/java/com/epam/test_generator/dto/UserDTO.java [hunk 2]
```diff
@@ -19,12 +20,15 @@
...
     private String password;
 
+    @NotNull
     private String role;
 
     private Integer attempts;
 
     private Boolean locked;
 
+    /*private List<ProjectDTO> projects;*/
```
Comment (artemSpb96) [inline src/main/java/com/epam/test_generator/dto/UserDTO.java]: Delete this, please.

File: src/main/java/com/epam/test_generator/entities/User.java [hunk 3]
```diff
@@ -24,6 +25,9 @@
 
     private Boolean locked;
 
+    /*@ManyToMany
+    private List<Project> projects;*/
+
```
Comment (artemSpb96) [inline src/main/java/com/epam/test_generator/entities/User.java]: And this :)

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java [hunk 4]
```diff
@@ -18,6 +20,28 @@
...
+        if (!project.isActive())
+            throw new ProjectClosedException("project is closed (readonly)");
+    }
+
+    public static void userBelongsToProject(Project project, User user) {
+        Set<User> userSet = project.getUsers();
+
+        if (userSet == null || userSet.stream()
+            .noneMatch(u -> Objects.equals(user.getId(), u.getId()))) {
+            throw new BadRequestException();
```
Comment (artemSpb96) [inline src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: Add message to exception, please.

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java [hunk 5]
```diff
@@ -18,6 +20,28 @@
...
+        }
+    }
+    public static void suitBelongsToProject(Project project, Suit suit) {
+        List<Suit> suitList = project.getSuits();
+
+        if (suitList == null || suitList.stream()
+            .noneMatch(s -> Objects.equals(s.getId(), suit.getId()))) {
+            throw new BadRequestException();
```
Comment (artemSpb96) [inline src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: And here.

File: src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import com.epam.test_generator.dto.ProjectDTO;
+import com.epam.test_generator.entities.Project;
+import com.epam.test_generator.entities.User;
+import java.util.List;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ProjectDAO extends JpaRepository<Project, Long> {
+    List<Project> findByUsers(User user);
```
[the remark this answers is not in the document]
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java]: It is correct

File: src/main/java/com/epam/test_generator/controllers/GlobalExceptionController.java [hunk 7]
```diff
@@ -6,21 +6,23 @@
...
 import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
-import org.springframework.security.access.AccessDeniedException;
 
 
 @ControllerAdvice
 public class GlobalExceptionController {
 
     @ExceptionHandler(Exception.class)
     public ResponseEntity<Void> handleRunTimeException(Exception ex) {
+        ex.printStackTrace();
```
[the remark this answers is not in the document]
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/controllers/GlobalExceptionController.java]: removed, tnx

File: src/main/java/com/epam/test_generator/controllers/StepController.java [hunk 8]
```diff
@@ -181,9 +198,10 @@
...
     @Secured({"ROLE_ADMIN", "ROLE_TEST_ENGINEER", "ROLE_TEST_LEAD"})
-    @RequestMapping(value = "/suits/{suitId}/cases/{caseId}/steps/{stepId}",
+    @RequestMapping(value = "/projects/{projectId}/suits/{suitId}/cases/{caseId}/steps/{stepId}",
         method = RequestMethod.DELETE)
-    public ResponseEntity<Void> removeCase(@PathVariable("suitId") long suitId,
+    public ResponseEntity<Void> removeCase(@PathVariable("projectId") long projectId,
+                                           @PathVariable("suitId") long suitId,
                                            @PathVariable("caseId") long caseId,
                                            @PathVariable("stepId") long stepId) {
         stepService.removeStep(suitId, caseId, stepId);
```
[the remark this answers is not in the document]
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/controllers/StepController.java]: Done, all controllers (suit, case, step, tag) are using projectId for validate now

File: src/main/java/com/epam/test_generator/services/ProjectService.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        AuthenticatedUser userDetails = (AuthenticatedUser) authentication.getPrincipal();
+        User authUser = userService.getUserByEmail(userDetails.getEmail());
+
+        Project project = projectTransformer.fromDto(projectDTO);
+        project.setUsers(Sets.newHashSet(authUser));
+        project.setActive(true);
+
+        projectDAO.save(project);
+
+        return project.getId();
```
[the remark this answers is not in the document]
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/services/ProjectService.java]: thank you! Fixed.

File: src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (artemSpb96) [reply src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java]: Yes, you are right)

File: src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java [hunk 10]
```diff
@@ -58,240 +59,240 @@ public void setup() {
...
-        doThrow(RuntimeException.class).when(suitService).updateSuit(anyLong(), any(SuitDTO.class));
+        doThrow(RuntimeException.class).when(suitService).updateSuit(anyLong(),anyLong(), any(SuitDTO.class));
 
-        mockMvc.perform(put("/suits/" + TEST_SUIT_ID)
+        mockMvc.perform(put("/projects/" + SIMPLE_PROJECT_ID + "/suits/" + TEST_SUIT_ID)
             .contentType(MediaType.APPLICATION_JSON)
             .content(mapper.writeValueAsString(suitDTO)))
             .andDo(print())
-            .andExpect(status().isInternalServerError());
+             .andExpect(status().isInternalServerError());
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java]: Space character breaks code formatting.

File: src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java [hunk 11]
```diff
@@ -58,240 +59,240 @@ public void setup() {
...
 
-        verify(suitService).updateSuit(anyLong(), any(SuitDTO.class));
+        verify(suitService).updateSuit(anyLong(),anyLong(), any(SuitDTO.class));
     }
 
     @Test
     public void removeSuit_return200whenRemoveSuit() throws Exception {
-        doNothing().when(suitService).removeSuit(anyLong());
+        doNothing().when(suitService).removeSuit(anyLong(),anyLong());
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java]: Possibly it slept off from here

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+            .collect(Collectors.toList());
+        expectedProjects = Stream.of(simpleProject1, simpleProject2)
+            .collect(Collectors.toList());
+    }
+
+    @Test
+    public void getProjects_expectedProjectsList() {
+        when(projectDAO.findAll()).thenReturn(expectedProjects);
+        final List<ProjectDTO> projects = projectService.getProjects();
+        assertEquals(projects, projects);
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: This assertion never fail

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+        simpleProject1.setUsers(userSet);
+        simpleProject2.setUsers(userSet);
+        userDTOSet = Stream.of(simpleUserDTO1, simpleUserDTO2)
+            .collect(Collectors.toSet());
+        simpleProject1DTO.setUsers(userDTOSet);
+        simpleProject2DTO.setUsers(userDTOSet);
+        simpleProjectFullDTO1.setUsers(userDTOSet);
+        expectedProjectDTOs = Stream.of(simpleProject1DTO, simpleProject2DTO)
+            .collect(Collectors.toList());
+        expectedProjects = Stream.of(simpleProject1, simpleProject2)
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: "expected" in test is name for parameter which goes first in assertion statement. Make sure expected used in that way.

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+        when(authentication.getPrincipal()).thenReturn(new AuthenticatedUser(
+            null,
+            simpleUser1.getEmail(),
+            null,
+            null,
+            false
+        ));
+        when(userService.getUserByEmail(simpleUser1.getEmail())).thenReturn(null);
+        final List<ProjectDTO> actualProjectDTOs = projectService
+            .getAuthUserProjects(authentication);
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: final List<ProjectDTO> actualProjectDTOs
this declaration can be removed.

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+        ));
+        when(projectDAO.findOne(simpleProject1.getId())).thenReturn(simpleProject1);
+        when(userService.getUserByEmail(simpleUser1.getEmail())).thenReturn(simpleUser1);
+        when(projectService.getProjectByProjectId(simpleProject1.getId())).thenReturn(simpleProject1);
+        ProjectFullDTO actualFullProject = projectService
+            .getAuthUserFullProject(simpleProjectFullDTO1.getId(), authentication);
+    }
+
+    @Test
+    public void createProject() {
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: Remove empty tests

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+        projectService.removeProject(SIMPLE_PROJECT_ID);
+        verify(projectDAO).delete(SIMPLE_PROJECT_ID);
+    }
+
+    @Test(expected = NotFoundException.class)
+    public void removeProject_invalidProjectId_NotFound() {
+
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        projectService.removeProject(777L);
+        verify(projectDAO).delete(SIMPLE_PROJECT_ID);
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: This line is never executed, is it?

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    @Test
+    public void addUserToProject_valid_Success() {
+        simpleProject1.setUsers(new HashSet<>());
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        when(userService.getUserById(SIMPLE_USER_ID)).thenReturn(simpleUser1);
+
+        assertFalse(simpleProject1.getUsers().contains(simpleUser1));
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: this assertion is pointless. It does not validate any piece of production code.

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+        projectService.addUserToProject(SIMPLE_PROJECT_ID, 666L);
+    }
+
+    @Test
+    public void removeUserFromProject_valid_Success() {
+        simpleProject1.setUsers(Sets.newHashSet(simpleUser1));
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        when(userService.getUserById(SIMPLE_USER_ID)).thenReturn(simpleUser1);
+
+        assertTrue(simpleProject1.getUsers().contains(simpleUser1));
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: This assertion does not test production code

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,398 @@
...
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        when(userService.getUserById(SIMPLE_USER_ID)).thenReturn(simpleUser1);
+
+        projectService.removeUserFromProject(SIMPLE_PROJECT_ID, 666L);
+    }
+
+    @Test
+    public void closeProject_validProjectId_Success() {
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        Project project = simpleProject1;
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: reassigning object makes test case hard to figure out, remove reassignment

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,398 @@
...
+        projectService.closeProject(SIMPLE_PROJECT_ID);
+        assertFalse(simpleProject1.isActive());
+        verify(projectDAO).save(project);
+    }
+
+    @Test(expected = NotFoundException.class)
+    public void closeProject_invalidProjectId_NotFound() {
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        projectService.closeProject(777L);
+        verify(projectDAO,never()).save(any(Project.class));
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: This line is never executed, is it?

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,398 @@
...
+    }
+
+    @Test(expected = ProjectClosedException.class)
+    public void closeProject_ProjectIsClosed_FailReadOnly() {
+        simpleProject1.setActive(false);
+        when(projectDAO.findOne(SIMPLE_PROJECT_ID)).thenReturn(simpleProject1);
+        projectService.closeProject(SIMPLE_PROJECT_ID);
+        verify(projectDAO,never()).save(any(Project.class));
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: This line is never executed, is it?

File: src/test/java/com/epam/test_generator/services/SuitServiceTest.java [hunk 22]
```diff
@@ -12,11 +13,13 @@
 import com.epam.test_generator.dao.interfaces.CaseVersionDAO;
 import com.epam.test_generator.dao.interfaces.SuitDAO;
 import com.epam.test_generator.dto.SuitDTO;
+import com.epam.test_generator.entities.Project;
 import com.epam.test_generator.entities.Suit;
 import com.epam.test_generator.services.exceptions.NotFoundException;
 import com.epam.test_generator.transformers.SuitTransformer;
 import java.util.ArrayList;
 import java.util.List;
+import org.hibernate.annotations.Check;
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/SuitServiceTest.java]: Unused import

File: src/test/java/com/epam/test_generator/services/SuitServiceTest.java [hunk 23]
```diff
@@ -109,22 +125,24 @@ public void updateSuitTest() {
...
         when(suitDAO.findOne(anyLong())).thenReturn(null);
 
-        suitService.updateSuit(SIMPLE_SUIT_ID, new SuitDTO());
+        suitService.updateSuit(SIMPLE_PROJECT_ID, SIMPLE_SUIT_ID, new SuitDTO());
     }
 
     @Test
     public void removeSuitTest() {
+        when(projectService.getProjectByProjectId(anyLong())).thenReturn(expectedProject);
+        when(projectService.getProjectByProjectId(anyLong())).thenReturn(expectedProject);
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/SuitServiceTest.java]: this line duplicates previous

File: src/test/java/com/epam/test_generator/services/SuitServiceTest.java [hunk 24]
```diff
@@ -136,7 +154,62 @@ public void removeSuitTest() {
...
+    }
+
+    @Test (expected = NotFoundException.class)
+    public void getSuitsFromProject_inValidId_NotFound() {
+        when(projectService.getProjectByProjectId(SIMPLE_PROJECT_ID)).thenReturn(expectedProject);
+        suitService.getSuitsFromProject(777L);
     }
 
+
+//    @Test
```
Comment (hdghg) [inline src/test/java/com/epam/test_generator/services/SuitServiceTest.java]: Remove commented code

File: src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (hdghg) [reply src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java]: Can it work if method name `findByUser`?

File: src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java [hunk 25]
```diff
@@ -0,0 +1,127 @@
+package com.epam.test_generator.dto;
+
+import com.epam.test_generator.entities.Suit;
+import com.epam.test_generator.entities.User;
+import java.util.List;
+import java.util.Set;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
+public class ProjectFullDTO {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java]: Provide javadoc explaining why second DTO is being added

File: src/main/java/com/epam/test_generator/entities/Project.java [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+    @Column(unique = true)
+    private String name;
+
+    private String description;
+
+    @OneToMany(cascade = {CascadeType.ALL})
+    private List<Suit> suits;
+
+    @ManyToMany
+    private Set<User> users;
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/entities/Project.java]: I do believe `participants` is more appropriate here.

File: src/main/java/com/epam/test_generator/entities/Project.java [hunk 27]
```diff
@@ -0,0 +1,129 @@
+package com.epam.test_generator.entities;
+
+import java.util.Set;
+import javax.persistence.*;
+import java.util.List;
+import java.util.Objects;
+
+@Entity
+public class Project {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/entities/Project.java]: Provide javadoc for this entity

File: src/main/java/com/epam/test_generator/services/ProjectService.java [hunk 28]
```diff
@@ -0,0 +1,129 @@
...
+import com.google.common.collect.Sets;
+import java.util.List;
+import javax.transaction.Transactional;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+
+@Service
+@Transactional
+public class ProjectService {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/services/ProjectService.java]: Provide javadoc here

File: src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java (hunk 1 above)
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java]: Added

File: src/main/java/com/epam/test_generator/dto/UserDTO.java (hunk 2 above)
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/dto/UserDTO.java]: Deleted

File: src/main/java/com/epam/test_generator/services/ProjectService.java [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+    private ProjectFullTransformer projectFullTransformer;
+
+    @Autowired
+    private UserService userService;
+
+    public List<ProjectDTO> getProjects() {
+        return projectTransformer.toDtoList(projectDAO.findAll());
+    }
+
+    public List<ProjectDTO> getAuthUserProjects(Authentication authentication) {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/services/ProjectService.java]: Avoid using abbreviations like `Auth`

File: src/main/java/com/epam/test_generator/entities/User.java (hunk 3 above)
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/entities/User.java]: Deleted

File: src/main/java/com/epam/test_generator/services/ProjectService.java (hunk 29 above)
Comment (hdghg) [reply src/main/java/com/epam/test_generator/services/ProjectService.java]: And add javadoc

File: src/main/java/com/epam/test_generator/services/exceptions/ProjectClosedException.java [hunk 30]
```diff
@@ -0,0 +1,11 @@
+package com.epam.test_generator.services.exceptions;
+
+public class ProjectClosedException extends RuntimeException {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/services/exceptions/ProjectClosedException.java]: Provide javadoc on how to use this exception

Comment (hdghg) [reply src/main/java/com/epam/test_generator/services/exceptions/ProjectClosedException.java]: When to throw and when to catch. And when caught, how to exctract additional data from it i.e. name of project that have already been closed.

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java [hunk 31]
```diff
@@ -18,6 +20,28 @@
             throw new NotFoundException();
     }
 
+    public static void checkProjectIsActive(Project project) {
+        if (!project.isActive())
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: Please use curly brackets in if-statements

File: src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (artemSpb96) [reply src/main/java/com/epam/test_generator/dao/interfaces/ProjectDAO.java]: no, spring data use property set in Project entity

File: src/main/java/com/epam/test_generator/controllers/GlobalExceptionController.java [hunk 32]
```diff
@@ -36,12 +38,16 @@
...
 
-
     @ExceptionHandler(BadRequestException.class)
     public ResponseEntity<Void> handleBadRequestException(BadRequestException ex) {
         return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
     }
 
+    @ExceptionHandler(ProjectClosedException.class)
+    public ResponseEntity<String> handleBadRequestException(ProjectClosedException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.FORBIDDEN);
```
[the remark this answers is not in the document]
Comment (PharaonSP) [reply src/main/java/com/epam/test_generator/controllers/GlobalExceptionController.java]: Done

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java (hunk 4 above)
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: fixed

File: src/main/java/com/epam/test_generator/services/ProjectService.java [hunk 33]
```diff
@@ -0,0 +1,129 @@
...
+
+        return project.getId();
+    }
+
+    public void updateProject(Long projectId, ProjectDTO projectDTO) {
+        Project project = projectDAO.findOne(projectId);
+        checkNotNull(project);
+        checkProjectIsActive(project);
+
+        projectTransformer.mapDTOToEntity(projectDTO, project);
```
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/ProjectService.java]: fixed

File: src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java [hunk 34]
```diff
@@ -0,0 +1,127 @@
+package com.epam.test_generator.dto;
+
+import com.epam.test_generator.entities.Suit;
+import com.epam.test_generator.entities.User;
+import java.util.List;
+import java.util.Set;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
```
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java]: added

File: src/main/java/com/epam/test_generator/entities/Project.java [hunk 35]
```diff
@@ -0,0 +1,129 @@
+package com.epam.test_generator.entities;
+
+import java.util.Set;
+import javax.persistence.*;
+import java.util.List;
+import java.util.Objects;
+
+@Entity
```
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/entities/Project.java]: done

File: src/main/java/com/epam/test_generator/entities/Project.java (hunk 27 above)
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/entities/Project.java]: added

File: src/main/java/com/epam/test_generator/services/ProjectService.java (hunk 29 above)
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/ProjectService.java]: done

[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/ProjectService.java]: added

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: done

File: src/main/java/com/epam/test_generator/services/exceptions/ProjectClosedException.java (hunk 30 above)
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/exceptions/ProjectClosedException.java]: fixed

File: src/main/java/com/epam/test_generator/services/utils/UtilsService.java (hunk 31 above)
Comment (AfanasievYuriy) [reply src/main/java/com/epam/test_generator/services/utils/UtilsService.java]: done

File: src/test/java/com/epam/test_generator/controllers/ProjectControllerTest.java [hunk 36]
```diff
@@ -0,0 +1,276 @@
...
+            .param("userId", "0"))
+            .andDo(print())
+            .andExpect(status().isForbidden());
+
+        verify(projectService).addUserToProject(anyLong(), anyLong());
+    }
+
+    @Test
+    public void removeUserToProject_ValidInput_Status200UserRemovedFromProject() throws Exception {
+        /*doThrow(ProjectClosedException.class)
```
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/controllers/ProjectControllerTest.java]: fixed (thanks)

File: src/test/java/com/epam/test_generator/controllers/ProjectControllerTest.java [hunk 37]
```diff
@@ -0,0 +1,276 @@
...
+        mockMvc.perform(delete("/projects/" + SIMPLE_PROJECT_ID +"/users")
+            .param("userId", "0"))
+            .andDo(print())
+            .andExpect(status().isNotFound());
+
+        verify(projectService).removeUserFromProject(anyLong(), anyLong());
+    }
+
+    @Test
+    public void removeUserToProject_ValidInput_Status403ProjectIsReadOnly() throws Exception {
```
[the remark this answers is not in the document]
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/controllers/ProjectControllerTest.java]: fixed

File: src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java (hunk 10 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java]: formatting (Ctrl Alt L) has been performed - thanks

File: src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java (hunk 11 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/controllers/SuitControllerTest.java]: fixed (was it about space char?)

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 13 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: ok (not for the first tome this happens)
will be more careful next time
p.s fixed

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 12 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: fixed (your message about "expected params" above)

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 14 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: removed (and in other places too)

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 15 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: fixed

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 16 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: fixed

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 19 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: removed

File: src/test/java/com/epam/test_generator/services/SuitServiceTest.java (hunk 23 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/SuitServiceTest.java]: removed

File: src/test/java/com/epam/test_generator/services/SuitServiceTest.java (hunk 22 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/SuitServiceTest.java]: removed

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 20 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: fixed (but didnt quite understand whats the problem)

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 21 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: fixed (but didnt quite understand whats the problem)

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 18 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: Why is it not? we mock projectDAO#findOne and userService#getUserById methods and projectDAO#save - they do all the stuff; this method just checks all data is ok and then add user to list of Projects. Basically we checks that simple project now contains simple user (but before execution of #addUserToProject) it was not

File: src/test/java/com/epam/test_generator/services/ProjectServiceTest.java (hunk 17 above)
Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: we use it to prove that simpleProject1 doesnt contain any users (especially simpleUser1) before 
#addUserToProject invokes. (more explonation in comment bellow)

Comment (AfanasievYuriy) [reply src/test/java/com/epam/test_generator/services/ProjectServiceTest.java]: or did you mean that we dont need  #assertFalse? (i thougth you mean whole test method is general)

File: src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java [hunk 38]
```diff
@@ -0,0 +1,141 @@
...
+/**
+ * DTO class, that fully represents {@link Project} entity (all it's fields). This DTO is used only for
+ * retrieving Projects from API (see GET method in {@link ProjectController}), because there is no need
+ * in container-fields (such as list of {@link Suit} with inner list of {@link Case} etc) for other
+ * methods (PUT, POST, DELETE) - in this cases usage of {@link ProjectDTO} is more convenient for
+ * API's users.
+ */
+public class ProjectFullDTO {
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/dto/ProjectFullDTO.java]: Please provide when to use ProjectDTO and when to use ProjectFullDTO

File: src/main/java/com/epam/test_generator/entities/Project.java [hunk 39]
```diff
@@ -0,0 +1,145 @@
+package com.epam.test_generator.entities;
+
+import java.util.HashSet;
+import java.util.Set;
+import javax.persistence.*;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This class represents Project essence. Besides simple fields like id, name and description,
```
Comment (hdghg) [inline src/main/java/com/epam/test_generator/entities/Project.java]: Please use correct translation. Entity, not essence.

**Pull request merged**